packages/flutter_tools/lib/src/android/java.dart

```diff
@@ -0,0 +1,245 @@
+  Map<String, String> getJavaEnvironment() {
+    return <String, String>{
+      if (home != null) _javaHomeEnvironmentVariable: home!,
+      'PATH': _fileSystem.path.dirname(binary) +
+                        _os.pathVarSeparator +
+                        _platform.environment['PATH']!,
+    };
+  }
+
+  /// Returns the version of java in the format \d(.\d)+(.\d)+
+  /// Returns null if version could not be determined.
+  JavaVersion? getVersion() {
+    if (_version == null) {
+      final RunResult result = _processUtils.runSync(
+        <String>[binary, '--version'],
+        environment: getJavaEnvironment(),
+      );
+      if (result.exitCode != 0) {
+        _logger.printTrace('java --version failed: exitCode: ${result.exitCode}'
+          ' stdout: ${result.stdout} stderr: ${result.stderr}');
+      }
+      _version = JavaVersion.tryParseFromJavaOutput(result.stdout, logger: _logger);
+    }
+
+    return _version;
+  }
+
+  bool canRun() {
+    return _processManager.canRun(binary);
+  }
+}
+
+String? _findJavaHome({
+  required Logger logger,
+  required AndroidStudio? androidStudio,
+  required Platform platform,
+}) {
+  final String? androidStudioJavaPath = androidStudio?.javaPath;
+  if (androidStudioJavaPath != null) {
+    logger.printTrace("Using Android Studio's java.");
+    return androidStudioJavaPath;
+  }
+
+  final String? javaHomeEnv = platform.environment[_javaHomeEnvironmentVariable];
+  if (javaHomeEnv != null) {
+    logger.printTrace('Using JAVA_HOME from environment valuables.');
+    return javaHomeEnv;
+  }
+  return null;
+}
+
+String? _findJavaBinary({
+  required Logger logger,
+  required String? javaHome,
+  required FileSystem fileSystem,
+  required OperatingSystemUtils operatingSystemUtils,
+  required Platform platform,
+}) {
+  try {
+    if (javaHome != null) {
+      return fileSystem.path.join(javaHome, 'bin', 'java');
+    }
+
+    // Fallback to PATH based lookup.
+    final String? pathJava = operatingSystemUtils.which(_kJavaExecutable)?.path;
+    if (pathJava != null) {
+      logger.printTrace('Using java from PATH.');
+    } else {
+      logger.printTrace('Could not find java path.');
+    }
+    return pathJava;
+  } on Exception catch (e) {
+    logger.printTrace('Exception thrown while searching for java:');
+    logger.printTrace(e.toString());
+    return null;
+  }
+}
+
+// Returns a user visible String that says the tool failed to parse
+// the version of java along with the output.
+String _formatJavaVersionWarning(String javaVersionRaw) {
+return 'Could not parse java version from: \n'
+    '$javaVersionRaw \n'
+    'If there is a version please look for an existing bug '
+    'https://github.com/flutter/flutter/issues/'
+    ' and if one does not exist file a new issue.';
+}
+
+class JavaVersion {
+  JavaVersion({
+    required this.longText,
+    required this.shortText
+  });
+
+  final String longText;
```

Comment:
can you add dartdoc's for `longText` and `shortText`?

Reply:
Done. Also renamed `shortText` to `number`.